File: SkyEditor.Core (VB)/TestComponents/MemoryConsoleProvider.vb
Imports System.Text
Imports SkyEditor.Core.ConsoleCommands

Namespace TestComponents
    Public Class MemoryConsoleProvider
        Implements IConsoleProvider

        Public Sub New()
            BackgroundColor = ConsoleColor.Black
            ForegroundColor = ConsoleColor.White
            StdIn = New StringBuilder
            StdOut = New StringBuilder
        End Sub

        Public Property BackgroundColor As ConsoleColor Implements IConsoleProvider.BackgroundColor

        Public Property ForegroundColor As ConsoleColor Implements IConsoleProvider.ForegroundColor

        Public Property StdIn As StringBuilder
        Private _stdInLock As New Object

        Public Property StdOut As StringBuilder
        Private _stdOutLock As New Object

        Public Sub Write(value() As Char) Implements IConsoleProvider.Write
            SyncLock _stdOutLock
                For Each c In value
                    StdOut.Append(c)
                Next
            End SyncLock
        End Sub

        Public Sub Write(value As String) Implements IConsoleProvider.Write
            SyncLock _stdOutLock
                StdOut.Append(value)
            End SyncLock
        End Sub

        Public Sub Write(value As Object) Implements IConsoleProvider.Write
            SyncLock _stdOutLock
                StdOut.Append(value.ToString)
            End SyncLock
        End Sub

        Public Sub Write(value As Boolean) Implements IConsoleProvider.Write
            SyncLock _stdOutLock
                StdOut.Append(value.ToString)
            End SyncLock
        End Sub

        Public Sub Write(format As String, ParamArray arg() As Object) Implements IConsoleProvider.Write
            SyncLock _stdOutLock
                StdOut.AppendFormat(format, arg)
            End SyncLock
        End Sub

        Public Sub Write(value() As Char, index As Integer, count As Integer) Implements IConsoleProvider.Write
            SyncLock _stdOutLock
                StdOut.Append(value, index, count)
            End SyncLock
        End Sub

        Public Sub WriteLine() Implements IConsoleProvider.WriteLine
            SyncLock _stdOutLock
                StdOut.AppendLine()
            End SyncLock
        End Sub

        Public Sub WriteLine(value() As Char) Implements IConsoleProvider.WriteLine
            SyncLock _stdOutLock
                For Each c In value
                    StdOut.Append(c)
                Next
                StdOut.AppendLine()
            End SyncLock
        End Sub

        Public Sub WriteLine(value As Object) Implements IConsoleProvider.WriteLine
            SyncLock _stdOutLock
                StdOut.AppendLine(value.ToString)
            End SyncLock
        End Sub

        Public Sub WriteLine(value As String) Implements IConsoleProvider.WriteLine
            SyncLock _stdOutLock
                StdOut.AppendLine(value)
            End SyncLock
        End Sub

        Public Sub WriteLine(format As String, ParamArray arg() As Object) Implements IConsoleProvider.WriteLine
            SyncLock _stdOutLock
                StdOut.AppendFormat(format, arg)
                StdOut.AppendLine()
            End SyncLock
        End Sub

        Public Sub WriteLine(value() As Char, index As Integer, count As Integer) Implements IConsoleProvider.WriteLine
            SyncLock _stdOutLock
                StdOut.Append(value, index, count)
                StdOut.AppendLine()
            End SyncLock
        End Sub

        Public Function Read() As Integer Implements IConsoleProvider.Read
            Dim c As Integer
            SyncLock _stdInLock
                If StdIn.Length > 0 Then
                    c = Convert.ToInt32(StdIn(0))
                    StdIn.Remove(0, 1)
                Else
                    c = -1
                End If
            End SyncLock
            Return c
        End Function

        Public Function ReadLine() As String Implements IConsoleProvider.ReadLine
            Dim line As New StringBuilder
            SyncLock _stdInLock
                While StdIn.Length > 0 AndAlso (line.Length = 0 OrElse line(line.Length - 1) <> vbLf)
                    line.Append(StdIn(0))
                    StdIn.Remove(0, 1)
                End While
            End SyncLock
            Return line.ToString.Trim
        End Function

        Public Function GetStdOut() As String
            SyncLock _stdOutLock
                Return StdOut.ToString
            End SyncLock
        End Function

    End Class
End Namespace

File: SkyEditor.Core/TestComponents/MemoryConsoleProvider.cs
using System;
using System.Collections.Generic;
using System.Text;
using SkyEditor.Core.ConsoleCommands;

namespace SkyEditor.Core.TestComponents
{
    public class MemoryConsoleProvider : IConsoleProvider
    {

        public MemoryConsoleProvider()
        {
            BackgroundColor = ConsoleColor.Black;
            ForegroundColor = ConsoleColor.White;
            StdIn = new StringBuilder();
            StdOut = new StringBuilder();
        }

        public ConsoleColor BackgroundColor { get; set; }

        public ConsoleColor ForegroundColor { get; set; }

        public StringBuilder StdIn { get; set; }

        private object _stdInLock = new object();
        public StringBuilder StdOut { get; set; }

        private object _stdOutLock = new object();
        public void Write(char[] value)
        {
            lock (_stdOutLock)
            {
                foreach (var c in value)
                {
                    StdOut.Append(c);
                }
            }
        }

        public void Write(string value)
        {
            lock (_stdOutLock)
            {
                StdOut.Append(value);
            }
        }

        public void Write(object value)
        {
            lock (_stdOutLock)
            {
                StdOut.Append(value.ToString());
            }
        }

        public void Write(bool value)
        {
            lock (_stdOutLock)
            {
                StdOut.Append(value.ToString());
            }
        }

        public void Write(string format, params object[] arg)
        {
            lock (_stdOutLock)
            {
                StdOut.AppendFormat(format, arg);
            }
        }

        public void Write(char[] value, int index, int count)
        {
            lock (_stdOutLock)
            {
                StdOut.Append(value, index, count);
            }
        }

        public void WriteLine()
        {
            lock (_stdOutLock)
            {
                StdOut.Append(Environment.NewLine);
            }
        }

        public void WriteLine(char[] value)
        {
            lock (_stdOutLock)
            {
                foreach (var c in value)
                {
                    StdOut.Append(c);
                }
                StdOut.Append(Environment.NewLine);
            }
        }

        public void WriteLine(object value)
        {
            lock (_stdOutLock)
            {
                StdOut.Append(value.ToString());
                StdOut.Append(Environment.NewLine);
            }
        }

        public void WriteLine(string value)
        {
            lock (_stdOutLock)
            {
                StdOut.Append(value);
                StdOut.Append(Environment.NewLine);
            }
        }

        public void WriteLine(string format, params object[] arg)
        {
            lock (_stdOutLock)
            {
                StdOut.AppendFormat(format, arg);
                StdOut.Append(Environment.NewLine);
            }
        }

        public void WriteLine(char[] value, int index, int count)
        {
            lock (_stdOutLock)
            {
                StdOut.Append(value, index, count);
                StdOut.Append(Environment.NewLine);
            }
        }

        public int Read()
        {
            char? c;
            lock (_stdInLock)
            {
                if (StdIn.Length > 0)
                {
                    c = StdIn[0];
                    StdIn.Remove(0, 1);
                }
                else
                {
                    c = null;
                }
            }
            return c.Value;
        }

        public string ReadLine()
        {
            if (StdIn.Length == 0)
            {
                // End of input
                return null;
            }

            StringBuilder line = new StringBuilder();
            lock (_stdInLock)
            {
                while (StdIn.Length > 0 && (line.Length == 0 || line[line.Length - 1] != '\n'))
                {
                    line.Append(StdIn[0]);
                    StdIn.Remove(0, 1);
                }
            }
            return line.ToString().Trim();
        }

        public string GetStdOut()
        {
            lock (_stdOutLock)
            {
                return StdOut.ToString();
            }
        }

    }
}
